### ISSUE TITLE
[Question] Downlink message were lost when thingsboard Edge disconnected.

### ISSUE THREAD

**Issue by VictorLiang-5giotlead** [thingsboard/thingsboard#8165]
**Component**
 * Rule Engine

**Description**
Downlink message were lost when ThingsBoard Edge instance active status is still true (server attribute - active: true) but network disconnection was actually happened in the edge.

**Question**
Is there any way to get a hint / notification if Edge did not receive this downlink or do something to prevent it?
Thanks for help!

**Environment**
 * OS:  Linux ( Debian 11 )
 * ThingsBoard: 3.4.4 CE
 * ThingsBoard Edge: 3.4.3EDGE
 * Browser: Mircosoft Edge 110.0.1587.57
 * ThingsBoard & ThingsBoard Edge deployed on different VM

**To Reproduce**
* my custom rule chain with default SequentialByOriginator strategy
  ![1](https://user-images.githubusercontent.com/97156419/222651413-2244b9bc-bfa8-424e-a82d-c2901678e187.png)
  ![2](https://user-images.githubusercontent.com/97156419/222651465-43b1bad5-20e2-4d45-a2a8-ee035c5b74c6.png)
* assign a device with custom rule chain to edge
* check device existed on edge
  ![3](https://user-images.githubusercontent.com/97156419/222651501-ed245987-a467-4471-b32a-0cebec917596.png)

	* Scenario 1:
		1 - disconnect edge's network
		2 - check edge's active attribute is still true
		3 - post a device shared attribute to create a downlink
![1-3](https://user-images.githubusercontent.com/97156419/222651699-c0b5ba57-b621-441b-b99b-fa95c48f6d39.png)
		4 - check downlink status ( should be Pending )         
![1-4](https://user-images.githubusercontent.com/97156419/222651829-045e4e4f-e624-40c0-8aea-90798a6ec63b.png)
		5 - tail ThingsBoard log and wait for 10 attempts
		```
		2023-03-03 04:30:35,165 [edge-send-scheduler-34-thread-1] WARN  2023-03-03 05:21:40,619 [edge-send-scheduler-34-thread-1] WARN  o.t.s.s.edge.rpc.EdgeGrpcSession - [9fbe0b74-e9c9-4849-86ac-666142652848] Failed to deliver the batch: [downlinkMsgId: 1328203544
		 entityData {
		   entityIdMSB: -5111817433467710995
		   entityIdLSB: -6042569554494185616
		   entityType: "DEVICE"
		   attributesUpdatedMsg {
			 kv {
			   key: "downlink_k"
			   type: STRING_V
			   string_v: "downlink_v"
			 }
		   }
		   postAttributeScope: "SHARED_SCOPE"
		 }
		 ], attempt: 10
		 2023-03-03 05:21:40,621 [edge-send-scheduler-34-thread-1] WARN  o.t.s.s.edge.rpc.EdgeGrpcSession - [9fbe0b74-e9c9-4849-86ac-666142652848] Failed to deliver the batch after 10 attempts. Next messages are going to be discarded [downlinkMsgId: 1328203544
		 entityData {
		   entityIdMSB: -5111817433467710995
		   entityIdLSB: -6042569554494185616
		   entityType: "DEVICE"
		   attributesUpdatedMsg {
			 kv {
			   key: "downlink_k"
			   type: STRING_V
			   string_v: "downlink_v"
			 }
		   }
		   postAttributeScope: "SHARED_SCOPE"
		 }
		 ]
		```
		6 - check downlink status ( should be Deployed now )
![1-6](https://user-images.githubusercontent.com/97156419/222652079-302790b9-fce6-4f76-aec4-db4e8f865cf7.png)
		7 - connect edge's network
		8 - check device's attribute
![1-8](https://user-images.githubusercontent.com/97156419/222652100-59a8fa5c-f7bd-4b47-be3b-d6436796c28a.png)

	* Scenario 2:
		1 - disconnect edge's network
		2 - check edge's active attribute is true
		3 - post a device shared attribute to create a downlink
![2-3](https://user-images.githubusercontent.com/97156419/222652219-5eda290e-a47b-4324-8cff-fb7a609d8229.png)
		4 - check downlink status ( should be Pending )
![2-4](https://user-images.githubusercontent.com/97156419/222652242-c9755ae7-66f5-44af-9cd7-d8d3dee29b41.png)
		5 - tail ThingsBoard log and connect edge's network before 10 attempts
		```
		2023-03-03 05:57:59,887 [edge-send-scheduler-34-thread-1] WARN  o.t.s.s.edge.rpc.EdgeGrpcSession - [8d952cc1-a6c8-4018-8c77-daca636f5227] Failed to deliver the batch: [downlinkMsgId: 1079316491
		entityData {
		  entityIdMSB: -5111817433467710995
		  entityIdLSB: -6042569554494185616
		  entityType: "DEVICE"
		  attributesUpdatedMsg {
			kv {
			  key: "downlink2_k"
			  type: STRING_V
			  string_v: "downlink2_v"
			}
		  }
		  postAttributeScope: "SHARED_SCOPE"
		}
		], attempt: 5
		2023-03-03 05:58:02,052 [grpc-default-executor-14] ERROR o.t.s.s.edge.rpc.EdgeGrpcSession - [8d952cc1-a6c8-4018-8c77-daca636f5227] Stream was terminated due to error:
		io.grpc.StatusRuntimeException: CANCELLED: client cancelled
				at io.grpc.Status.asRuntimeException(Status.java:526)
				at io.grpc.stub.ServerCalls$StreamingServerCallHandler$StreamingServerCallListener.onCancel(ServerCalls.java:291)
				at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.closedInternal(ServerCallImpl.java:362)
				at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.closed(ServerCallImpl.java:350)
				at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1Closed.runInContext(ServerImpl.java:913)
				at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
				at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
				at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
				at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
				at java.base/java.lang.Thread.run(Thread.java:829)
		2023-03-03 05:58:02,052 [grpc-default-executor-14] INFO  o.t.s.s.edge.rpc.EdgeGrpcService - [fd568920-b981-11ed-ac24-7b014dc89f70] edge disconnected!
		```
		6 - check downlink status ( should be Deployed now )
![2-6](https://user-images.githubusercontent.com/97156419/222652265-6e62b920-dff3-4fb4-a10d-85c1b29e09d8.png)
		7 - check edge's active attribute is false
![2-7](https://user-images.githubusercontent.com/97156419/222652286-80bc6379-2296-4b76-aa13-16c8c19a75d6.png)
		8 - check device's attribute
![2-8](https://user-images.githubusercontent.com/97156419/222652316-eb20eb49-5f98-4981-977c-904d1e037f00.png)

**Comment by star7128** [comment]
We had a similar issue like yours. Maybe we need TB edge team involved as well.

Briefly, we see downlink 10 times retries. 

In scenario 1, you wait until 10 times retries failure first. To recover edge network connection with TB.
In scenario 2, only few times retries and quickly recover edge network connection with TB.

However, you could not  see post a device shared attribute successfully no matter scenario 1 or 2.

I assume it might be a bug or some states are not defined very well. 

Anyway, let us see if anyone is able to dig more or look forward to hearing acknowledge from ThingsBoard team members.

**Comment by volodymyr-babak** [comment]
@VictorLiang-5giotlead @star7128 

Thanks for posting this issue. 
Default keep alive time for edge client is 360 seconds. 
According to your feedback, probably we'll need to decrease it to avoid having this type of issues.
Could you please check if it helps in your case, if you decrease this interval to 5 or 10 seconds. 
Could you please add next line into tb-edge.conf file and restart edge service to validate it:
`export CLOUD_RPC_KEEP_ALIVE_TIME_SEC=5`

Please let me know if this will fix your cases.

**Comment by VictorLiang-5giotlead** [comment]
> @VictorLiang-5giotlead @star7128
> 
> Thanks for posting this issue. Default keep alive time for edge client is 360 seconds. According to your feedback, probably we'll need to decrease it to avoid having this type of issues. Could you please check if it helps in your case, if you decrease this interval to 5 or 10 seconds. Could you please add next line into tb-edge.conf file and restart edge service to validate it: `export CLOUD_RPC_KEEP_ALIVE_TIME_SEC=5`
> 
> Please let me know if this will fix your cases.

Thanks for your reply.
I try to decrease edge's keep alive time to 5 seconds,
* Scenario 1
	* Get the same resault as Scenario 2 - edge attribute `active` changes to false before 10 attempts and downlink status chnage to `Deployed`, but edge did not receive this downlink.

* Scenario 2
	* Still happened.

here is ThingsBoard logs
```
2023-03-29 11:33:46,576 [edge-send-scheduler-31-thread-1] WARN  o.t.s.s.edge.rpc.EdgeGrpcSession - [cc75659a-549a-4be4-8246-c4a728b9cd13] Failed to deliver the batch: [downlinkMsgId: 1520481830
entityData {
  entityIdMSB: 6148948413047247341
  entityIdLSB: -7516053181382356113
  entityType: "DEVICE"
  attributesUpdatedMsg {
    kv {
      key: "downlink_k"
      type: STRING_V
      string_v: "downlink_v"
    }
  }
  postAttributeScope: "SHARED_SCOPE"
}
], attempt: 2
2023-03-29 11:33:49,162 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-1 [TS] queueSize [0] totalAdded [2] totalSaved [2] totalFailed [0]
2023-03-29 11:33:49,162 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-2 [TS] queueSize [0] totalAdded [2] totalSaved [2] totalFailed [0]
2023-03-29 11:33:49,219 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-1 [TS Latest] queueSize [0] totalAdded [2] totalSaved [2] totalFailed [0]
2023-03-29 11:33:49,219 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-2 [TS Latest] queueSize [0] totalAdded [2] totalSaved [2] totalFailed [0]
2023-03-29 11:33:55,522 [grpc-default-executor-8] ERROR o.t.s.s.edge.rpc.EdgeGrpcSession - [cc75659a-549a-4be4-8246-c4a728b9cd13] Stream was terminated due to error:
io.grpc.StatusRuntimeException: CANCELLED: client cancelled
	at io.grpc.Status.asRuntimeException(Status.java:526)
	at io.grpc.stub.ServerCalls$StreamingServerCallHandler$StreamingServerCallListener.onCancel(ServerCalls.java:291)
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.closedInternal(ServerCallImpl.java:362)
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.closed(ServerCallImpl.java:350)
	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1Closed.runInContext(ServerImpl.java:913)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
2023-03-29 11:33:55,522 [grpc-default-executor-8] INFO  o.t.s.s.edge.rpc.EdgeGrpcService - [219cbce0-cddd-11ed-a030-5b6e5d6c9c01] edge disconnected!
```

I think the problem is:

After an abnormal disconnect, such as Edge's network disconnect, the gRPC session can't detect it immediately and still alive for ThingsBoard before a messaging failure occurs. If any message is sent at this moment, the gRPC session will disconnect, but the other attempts will be dropped, and the messaging won't retry after the Edge connects again.

**Comment by volodymyr-babak** [comment]
@VictorLiang-5giotlead 

thanks for replying back with additional information.
We'll investigate it more deeply and provide you with updates.

**Comment by VictorLiang-5giotlead** [comment]
> @VictorLiang-5giotlead
> 
> thanks for replying back with additional information. We'll investigate it more deeply and provide you with updates.

Update:
It made me feel weird about the result yesterday, so I tried again with a brand new Thingsboard & Edge and set Edge's environment in docker-compose.yml `CLOUD_RPC_KEEP_ALIVE_TIME_SEC: 5`.

* Scenario 1:
Still happening and getting the same result as `default keep alive time: 360 seconds` - after 10 attempts, the message will be discarded and the downlink status will change to Deployed, but the Edge did not receive it.

ThingsBoard logs
```
in-memory-mytb-1  | 2023-03-30 04:08:05,838 [edge-send-scheduler-33-thread-1] WARN  o.t.s.s.edge.rpc.EdgeGrpcSession - [c0d37e30-f32e-4988-93df-8d26a60c5df7] Failed to deliver the batch: [downlinkMsgId: 861060978
in-memory-mytb-1  | entityData {
in-memory-mytb-1  |   entityIdMSB: -5015898031570218515
in-memory-mytb-1  |   entityIdLSB: -6007479625658596654
in-memory-mytb-1  |   entityType: "DEVICE"
in-memory-mytb-1  |   attributesUpdatedMsg {
in-memory-mytb-1  |     kv {
in-memory-mytb-1  |       key: "downlink_k"
in-memory-mytb-1  |       type: STRING_V
in-memory-mytb-1  |       string_v: "downlink_v"
in-memory-mytb-1  |     }
in-memory-mytb-1  |   }
in-memory-mytb-1  |   postAttributeScope: "SHARED_SCOPE"
in-memory-mytb-1  | }
in-memory-mytb-1  | ], attempt: 10
in-memory-mytb-1  | 2023-03-30 04:08:05,842 [edge-send-scheduler-33-thread-1] WARN  o.t.s.s.edge.rpc.EdgeGrpcSession - [c0d37e30-f32e-4988-93df-8d26a60c5df7] Failed to deliver the batch after 10 attempts. Next messages are going to be discarded [downlinkMsgId: 861060978
in-memory-mytb-1  | entityData {
in-memory-mytb-1  |   entityIdMSB: -5015898031570218515
in-memory-mytb-1  |   entityIdLSB: -6007479625658596654
in-memory-mytb-1  |   entityType: "DEVICE"
in-memory-mytb-1  |   attributesUpdatedMsg {
in-memory-mytb-1  |     kv {
in-memory-mytb-1  |       key: "downlink_k"
in-memory-mytb-1  |       type: STRING_V
in-memory-mytb-1  |       string_v: "downlink_v"
in-memory-mytb-1  |     }
in-memory-mytb-1  |   }
in-memory-mytb-1  |   postAttributeScope: "SHARED_SCOPE"
in-memory-mytb-1  | }
in-memory-mytb-1  | ]
```

* Scenario 2:
Almost fixed, it only happens a few times, and it's hard to reproduce.

Please let me know if you need any information about this issue.
Thank you and your team for this amazing platform!

**Comment by volodymyr-babak** [comment]
This issue is resolved by this commit https://github.com/thingsboard/thingsboard/pull/8346. 
Updated keep alive functionality will be available in 3.5 release.
